File: Python/125-Valid-Palindrome.py
class Solution:
  '''
  Two Pointer Approach
    * TC: O(n), where n is the length of string s
    * SC: O(1), no extra space is allocated
  '''
  def isPalindrome(self, s: str) -> bool:
    left = 0
    right = len(s) - 1

    while left < right:
      while left < right and not alphanum(s[left]): # checking to see if character is alphanumeric 
        left += 1
      
      while left < right and not alphanum(s[right]):# checking to see if character is alphanumeric 
        right -= 1
      
      if s[left].lower() != s[right].lower(): # checking to see if left and right chars are the same
        return False  # string s is not a palindrome
      
      left += 1
      right -= 1

    return True # string s is a palindrome

# checks for alphanumeric character c
def alphanum(c):
  return (
    ord("A") <= ord(c) <= ord("Z")
    or ord("a") <= ord(c) <= ord("z")
    or ord("0") <= ord(c) <= ord("9")
  )

File: C++/125-Valid-Palindrome.cpp
#include <iostream>
#include <cctype>
#include <string>
using namespace std;

class Solution {
  public:
    /**
     * Two Pointer Approach
     * TC: O(n), where n is the length of string s
     * SC: O(1), no extra space is allocated
     */
    bool isPalindrome(string s) {
      int left = 0;             // left pointer
      int right = s.size() - 1; // right pointer

      while (left < right) {
        while (!isalnum(s[left]) && left < right) { // checking to see if character is alphanumeric 
          left++;
        }
        while (!isalnum(s[right]) && left < right) {// checking to see if character is alphanumeric 
          right--;
        }
        if (tolower(s[left]) != tolower(s[right])) {// checking to see if left and right chars are the same
          return false; // string s is not a palindrome
        }
        left++;
        right--;
      }

      return true;  // string s is a palindrome
    }
};